universal-idl-program/program/src/instructions/create_idl/mod.rs

```diff
@@ -0,0 +1,208 @@
+    sysvar::{clock::Clock, rent::Rent, Sysvar},
+};
+
+pub fn create_idl(
+    program_id: Pubkey,
+    payer: Pubkey,
+    program_authority: Pubkey,
+    idl_account: Pubkey,
+    program: Pubkey,
+    program_data_or_frozen_auth: Pubkey,
+) -> Result<Instruction, ProgramError> {
+    Ok(Instruction {
+        program_id,
+        accounts: vec![
+            AccountMeta::new(payer, true),
+            AccountMeta::new_readonly(program_authority, true),
+            AccountMeta::new(idl_account, false),
+            AccountMeta::new_readonly(program, false),
+            AccountMeta::new_readonly(program_data_or_frozen_auth, false),
+            AccountMeta::new_readonly(solana_program::system_program::id(), false),
+        ],
+        data: IdlProgramInstruction::CreateIdl.try_to_vec()?,
+    })
+}
+
+pub enum AuthoritySource<'a, 'info> {
+    UpgradeableAuthority {
+        upgradeable_program: &'a AccountInfo<'info>,
+        program_data: UpgradeableLoaderState,
+    },
+    FrozenProgramAuthority {
+        frozen_program: &'a AccountInfo<'info>,
+        frozen_authority: FrozenProgramAuthority,
+    },
+}
+impl<'a, 'info> AuthoritySource<'a, 'info> {
+    pub fn check_authority(&self, authority: &Pubkey) -> ProgramResult {
+        match self {
+            Self::FrozenProgramAuthority {
+                frozen_authority, ..
+            } => assert_address(authority, &frozen_authority.authority, "Frozen authority"),
+            Self::UpgradeableAuthority { program_data, .. } => {
+                assert_program_authority_matches_program(program_data, authority)
+            }
+        }
+    }
+
+    pub fn load(
+        accounts_iter: &mut impl Iterator<Item = &'a AccountInfo<'info>>,
+    ) -> Result<Self, ProgramError> {
+        let program = next_account_info(accounts_iter)?;
+        assert_executable(program, "program")?;
+        if is_upgradeable_program(program) {
+            let program_data = next_account_info(accounts_iter)?;
+
+            // program
+            let program_state: UpgradeableLoaderState =
+                bincode::deserialize(&program.data.borrow())
+                    .map_err(|_| ProgramError::InvalidAccountData)?;
+            assert_program_data_matches_program(&program_state, program_data.key)?;
+
+            // program_data
+            assert_account_owned_by_loader(program_data, "program_data")?;
+            let program_data_state: UpgradeableLoaderState =
+                bincode::deserialize(&program_data.data.borrow())
+                    .map_err(|_| ProgramError::InvalidAccountData)?;
+
+            Ok(Self::UpgradeableAuthority {
+                upgradeable_program: program,
+                program_data: program_data_state,
+            })
+        } else if is_frozen_program(program) {
+            let frozen_authority_account = next_account_info(accounts_iter)?;
+            assert_frozen_authority_seeds(program.key, frozen_authority_account.key)?;
```

Comment:
I think this function is semantically difficult to reason about. I know it's checking something important, but it's unclear what it's doing.

If `frozen_authority_account` is a PDA and this is performing verification, I think it would be apt to either add a comment or rename the function `assert_correct_frozen_authority_address` or something of that nature.